### PR TITLE
186293095 - Add more connections dinamically

### PR DESCRIPTION
**Pull request by iduartgomez**
Changes in this PR:
- Peers will try to acquire more connections automatically now if they don't have reached the max number of connections.
 Now we have a background task that will maintain the quality and number of connections automatically.
- Ideal join location criteria selected by topology manager now.
- Operations now time out and are clean up automatically.
- Fixed multiple minor issues with the connect op so all state transitions are correct now.
- Fixed some issues with the event log serialization, more robust testing.
- Optimized transaction type so is more packed and faster to copy (only 16 bytes size now)

### CODE REVIEW

File: crates/core/src/ring.rs [hunk 1]
```diff
@@ -251,62 +381,65 @@ impl Ring {
...
-            // todo: in the future maybe use the `small worldness` metric to decide
-            let median_distance = self
-                .median_distance_to(my_location)
-                .unwrap_or(Distance(0.5));
-            let dist_to_loc = my_location.distance(location);
-            let is_lower_than_median = dist_to_loc < median_distance;
-            tracing::debug!("dist to connection loc: {dist_to_loc}, median dist: {median_distance}, accepting: {is_lower_than_median}");
-            is_lower_than_median
+            let strategy = if self
+                .fast_acquisition
```
Comment (sanity) [inline crates/core/src/ring.rs:403]: May be tricky, I realized that there is no one ideal small world topology, it depends on how big the network is. Small network won't have much clustering, large network will. I think the best metric of how good the topology is is whether outbound requests are evenly distributed among connections - so lower standard deviation is better.

File: crates/core/src/ring.rs [hunk 2]
```diff
@@ -251,62 +381,65 @@ impl Ring {
...
-            tracing::debug!(peer = %self.peer_key, "max connections reached");
+            tracing::debug!(peer = %self.peer_key, "max open connections reached");
             false
         } else {
-            // todo: in the future maybe use the `small worldness` metric to decide
-            let median_distance = self
-                .median_distance_to(my_location)
-                .unwrap_or(Distance(0.5));
-            let dist_to_loc = my_location.distance(location);
-            let is_lower_than_median = dist_to_loc < median_distance;
```
Comment (sanity) [inline crates/core/src/ring.rs:274]: What is the relevance of whether it's lower than the median?

File: crates/core/src/ring.rs [hunk 3]
```diff
@@ -431,12 +566,240 @@ impl Ring {
...
+            ) else {
+                return Ok(());
+            };
+            msg
+        };
+        notifier.send(Either::Left((msg.into(), None))).await?;
+        Ok(())
+    }
+
+    fn find_optimal_query_target<'a>(
```
Comment (sanity) [inline crates/core/src/ring.rs:764]: I'm unclear on the purpose of this, isn't this what the Router is for?

File: crates/core/src/ring.rs [hunk 4]
```diff
@@ -603,6 +986,14 @@ impl Display for Distance {
     }
 }
 
+impl std::ops::Sub<Distance> for Distance {
+    type Output = f64;
+
+    fn sub(self, rhs: Distance) -> Self::Output {
+        self.0 - rhs.0
```
Comment (sanity) [inline crates/core/src/ring.rs:993]: Could this result in a negative distance?

**Review by sanity — COMMENTED** [review]
A few questions

File: crates/core/src/ring.rs (hunk 4 above)
Comment (iduartgomez) [reply crates/core/src/ring.rs:993]: Yes, however this is not needed anymore so I am going to remove this completely.

File: crates/core/src/ring.rs (hunk 3 above)
Comment (iduartgomez) [reply crates/core/src/ring.rs:764]: I guess we could be using the router here yes. 

The goal is to send a request to join peer C from peer A to an optimal query target, peer B. Here the optimal was the closes to the ideal location which was an heuristic that was simple enough (after all, the target which is the closest to the ideal location, is likely to have the most connections to peers closest to the ideal location. Peer B here is just acting as a mere proxy to forward that request to the final receiver of the request so not sure if using the router would be necessary but I guess it won't hurt.

**Pull request merged**